ENIGMAsystem/SHELL/Universal_System/Serialization/serialization.h

```diff
@@ -0,0 +1,817 @@
+
+template <typename T>
+typename std::enable_if<std::is_base_of_v<variant, std::decay_t<T>>,
+                        std::vector<std::byte>>::type inline internal_serialize_fn(T &&value) {
+  return internal_serialize_variant(value);
+}
+
+template <typename T>
+typename std::enable_if<std::is_integral_v<std::decay_t<T>> || std::is_floating_point_v<std::decay_t<T>>,
+                        std::array<std::byte, sizeof(T)>>::type inline internal_serialize_fn(T &&value) {
+  return internal_serialize_numeric(value);
+}
+
+template <typename T>
+typename std::enable_if<is_std_vector_v<std::decay_t<T>>, std::vector<std::byte>>::type inline internal_serialize_fn(
+    T &&value) {
+  std::vector<std::byte> result;
+  result.resize(sizeof(std::size_t) + value.size() * ((value.size()) ? enigma_internal_sizeof(value[0]) : 0));
+  internal_serialize_into<std::size_t>(result.data(), value.size());
+
+  auto dataPtr = result.data() + sizeof(std::size_t);
+  for (const auto &element : value) {
+    internal_serialize_into(dataPtr, element);
+    dataPtr += enigma_internal_sizeof(element);
+  }
+  return result;
+}
+
+template <typename T>
+typename std::enable_if<is_std_map_v<std::decay_t<T>>, std::vector<std::byte>>::type inline internal_serialize_fn(
+    T &&value) {
+  std::vector<std::byte> result;
+  result.resize(sizeof(std::size_t) + value.size() * ((value.size()) ? (enigma_internal_sizeof(value.begin()->first) +
+                                                                        enigma_internal_sizeof(value.begin()->second))
+                                                                     : 0));
+  internal_serialize_into<std::size_t>(result.data(), value.size());
+
+  auto dataPtr = result.data() + sizeof(std::size_t);
+  for (const auto &element : value) {
+    internal_serialize_into(dataPtr, element.first);
+    internal_serialize_into(dataPtr, element.second);
+    dataPtr += enigma_internal_sizeof(element.first) + enigma_internal_sizeof(element.second);
+  }
+  return result;
+}
+
+template <typename T>
+typename std::enable_if<is_std_complex_v<std::decay_t<T>>, std::vector<std::byte>>::type inline internal_serialize_fn(
+    T &&value) {
+  std::vector<std::byte> result;
+  result.resize(2 * enigma_internal_sizeof(value.real()));
+
+  auto dataPtr = result.data();
+  internal_serialize_into(dataPtr, value.real());
+  internal_serialize_into(dataPtr, value.imag());
+  return result;
+}
+
+template <typename T>
+inline auto internal_serialize(T &&value) {
+  if constexpr (has_internal_serialize_fn_free_function<std::decay_t<T>>) {
+    return internal_serialize_fn(value);
+  } else if constexpr (has_serialize_method_v<std::decay_t<T>>) {
+    return value.serialize();
+  } else {
+    static_assert(always_false<T>,
+                  "'serialize' takes 'variant', 'var', 'std::string', bool, integral, floating types, std::vector, "
+                  "std::map or std::complex");
+  }
+}
```

Comment:
I think the most important thing you should do before adding functions for more types is to update all of these to use the const-reference parameters and pattern match instead of using `is_same_v` so that it is less effort to move them later.

Reply:
ok, I will do that.